invoices/invoiceregistry.go

```diff
@@ -1043,3 +1141,309 @@ func (i *InvoiceRegistry) HodlUnsubscribeAll(subscriber chan<- interface{}) {
+
+	case resultDuplicateToSettled:
+		return "accepting duplicate payment to settled invoice"
+
+	case resultAccepted:
+		return "accepted"
+
+	case resultSettled:
+		return "settled"
+
+	case resultInvoiceNotOpen:
+		return "invoice no longer open"
+
+	case resultPartialAccepted:
+		return "partial payment accepted"
+
+	case resultMppInProgress:
+		return "mpp reception in progress"
+
+	case resultAddressMismatch:
+		return "payment address mismatch"
+
+	case resultHtlcSetTotalMismatch:
+		return "htlc total amt doesn't match set total"
+
+	case resultHtlcSetTotalTooLow:
+		return "set total too low for invoice"
+
+	case resultHtlcSetOverpayment:
+		return "mpp is overpaying set total"
+
+	default:
+		return "unknown"
+	}
+}
+
+// invoiceUpdateCtx is an object that describes the context for the invoice
+// update to be carried out.
+type invoiceUpdateCtx struct {
+	circuitKey           channeldb.CircuitKey
+	amtPaid              lnwire.MilliSatoshi
+	expiry               uint32
+	currentHeight        int32
+	finalCltvRejectDelta int32
+	mpp                  *record.MPP
+}
+
+// updateInvoice is a callback for DB.UpdateInvoice that contains the invoice
+// settlement logic.
+func updateInvoice(ctx *invoiceUpdateCtx, inv *channeldb.Invoice) (
+	*channeldb.InvoiceUpdateDesc, updateResult, error) {
+
+	// Don't update the invoice when this is a replayed htlc.
+	htlc, ok := inv.Htlcs[ctx.circuitKey]
+	if ok {
+		switch htlc.State {
+		case channeldb.HtlcStateCanceled:
+			return nil, resultReplayToCanceled, nil
+
+		case channeldb.HtlcStateAccepted:
+			return nil, resultReplayToAccepted, nil
+
+		case channeldb.HtlcStateSettled:
+			return nil, resultReplayToSettled, nil
+
+		default:
+			return nil, 0, errors.New("unknown htlc state")
+		}
+	}
+
+	if ctx.mpp == nil {
+		return updateLegacy(ctx, inv)
+	}
+
+	return updateMpp(ctx, inv)
+}
+
+// updateMpp is a callback for DB.UpdateInvoice that contains the invoice
+// settlement logic for mpp payments.
+func updateMpp(ctx *invoiceUpdateCtx, inv *channeldb.Invoice) (
+	*channeldb.InvoiceUpdateDesc, updateResult, error) {
+
+	// Start building the accept descriptor.
+	acceptDesc := &channeldb.HtlcAcceptDesc{
+		Amt:          ctx.amtPaid,
+		Expiry:       ctx.expiry,
+		AcceptHeight: ctx.currentHeight,
+		MppTotalAmt:  ctx.mpp.TotalMsat(),
+	}
+
+	// Only accept payments to open invoices. This behaviour
+	// differs from non-mpp payments that are accepted even
+	// after the invoice is settled. Because non-mpp
+	// payments don't have a payment address, this is needed
+	// to thwart probing.
+	if inv.State != channeldb.ContractOpen {
+		return nil, resultInvoiceNotOpen, nil
+	}
+
+	// Check the payment address that authorizes the
+	// payment.
+	if ctx.mpp.PaymentAddr() != inv.Terms.PaymentAddr {
+		return nil, resultAddressMismatch, nil
+	}
+
+	// If there is a set invoice amount, check that the
+	// total amt of the htlc set is high enough.
+	if inv.Terms.Value > 0 {
```

Comment:
is this needed? if `inv.Terms.Value == 0`, then it should still procude the same result if `ctx.mpp.TotalMsat < inv.Terms.Value` is checked directly. also handles cases where `TotalMsat()` is interpreted as negative

Reply:
Yes, clause removed. Also added an additional check for set_amt==0